apps/openmw/mwmechanics/spellcasting.cpp

```diff
@@ -763,6 +763,35 @@ namespace MWMechanics
 
         int school = 0;
 
+        if(mCaster.getClass().isActor()
```

Comment:
The castSpell function isn't the correct place to do this check. There are other ways of casting effects (enchantments, potions, etc.) that don't go through castSpell.

Either you do it in somewhere in CastSpell::inflict, or directly in PhysicsSystem where Water Walking is actually applied (preferable IMO, since you can easily get actor's physical height from there).

Observation in MW: The underwater check isn't done when the spell is cast, but rather every frame. You can test this by disabling collision (TCL), casting water walking, going underwater, then do another TCL (player is moved above water). The implementation in physics system would support this use case.